sympy/solvers/simplex.py

```diff
@@ -0,0 +1,821 @@
+    """return A, B, C, D, r, x+X, X for maximizing
+    objective = Cx - D with constraints Ax <= B, introducing
+    introducing auxilliary variables, X, as necessary to make
+    replacements of symbols as given in r, {xi: expression with Xj},
+    so all variables in x+X will take on nonnegative values.
+
+    If symbols not in objective will be set to 0.
+    """
+
+    # sympify input and collect free symbols
+    f = sympify(objective)
+    syms = f.free_symbols
+    constraints = [sympify(i) for i in constraints]
+
+    # set to 0 any symbol not in the objective: it's
+    # value does not matter
+    cfree = set()
+    for i in constraints:
+        cfree |= i.free_symbols
+    zero = dict(zip(cfree - syms, [0]*len(cfree)))
+    constraints = [i.xreplace(zero) for i in constraints]
+
+    # convert constraints to nonpositive expressions
+    _ = _rel_as_nonpos(constraints)
+    if _ is None:
+        raise InfeasibleLPError(
+            'inconsistent/False constraint')
+    np, r, aux = _
+
+    # do change of variables
+    F = f.xreplace(r)
+    np = [i.xreplace(r) for i in np]
+
+    # convert to matrices
+    xx = list(ordered(syms)) + aux
+    A, B = linear_eq_to_matrix(np, xx)
+    C, D = linear_eq_to_matrix([F], xx)
+    return A, B, C, D, r, xx, aux
+
+
+def _lp(min_max, f, constr):
+    """Return the optimization (min or max) of ``f`` with the given
+    constraints. Unless constrained, variables will assume only
+    nonnegative values in the solution. A constraint like ``x <= 2``
+    will permit ``x`` to have negative values while ``x < oo`` will
+    allow it to have any real value.
+
+    If `min_max` is 'max' then the results corresponding to the
+    maximization of ``f`` will be returned, else the minimization.
+    The constraints can be given as Le, Ge or Eq expressions.
+
+    Examples
+    ========
+
+    >>> from sympy.solvers.simplex import _lp as lp
+    >>> from sympy import symbols, Eq, oo
+    >>> from sympy.abc import x, y
+    >>> x1, x2, x3, x4 = symbols('x1:5')
+    >>> f = 5*x2 + x3 + 4*x4 - x1
+    >>> c = [x1 + 5 >= 5*x2 + 2*x3 + 5*x4,
+    ...      Eq(3*x2 + x4, 2),
+    ...      Eq(-x1 + x3 + 2*x4, 1)]
+    >>> lp(min, f, c)
+    (9/2, {x1: 0, x2: 1/2, x3: 0, x4: 1/2})
+
+    By passing max, the maximum value for f under the constraints
+    is returned if possible:
+
+    >>> lp(max, f, c + [x3 < oo])  # x3 can have any value
+    (5, {x1: 1, x2: 0, x3: -2, x4: 2})
+
+    Symbols not in the objective function will not be reported
+    and are treated as 0:
+
+    >>> lp(min, x, [y + x >= -3])  # same as lp(min, x, [x >= -3])
+    (-3, {x: -3})
+    """
+    A, B, C, D, r, xx, aux = _lp_matrices(f, constr)
+
+    how = str(min_max).lower()
+    if 'max' in how:
+        # _simplex minimizes for Ax <= B so we
+        # have to change the sign of the function
+        # and negate the optimal value returned
+        _o, p, d = _simplex(A, B, -C, -D)
+        o = -_o
```

Comment:
Are you sure this works?

I would think that flipping the sign of the objective function and then flipping the sign of the optimum value wouldn't work because of the constraints.

Wouldn't something along these lines happen?

maximize x + y subject to 
x <= 2
y <= 3

becomes

minimize x + y
subject to 
x <= 2
y <= 3

which is unbounded even though the original problem was not.

Reply:
Oh I see _lp_args handles this I think.

Reply:
I see that it's necessary to keep min_max as an input to this function so that the sign of the final output can be flipped. Couldn't the flipping of the sign of f take place in _lp_args?

Reply:
I think it might be better to handle all of the logic for converting the maximization problem to a minimization problem and back within `lpmax`. 

The reduction from maximization problem to minimization problem should be very clear.

Reply:
To make the reduction clear I think it should take place together, not split up over multiple functions.

Reply:
> Couldn't the flipping of the sign of f take place in _lp_args?

`_lp_args` returns f and const in standard from from the matrices. It is only because `_simplex` doesn't solve the standard form that the handling specifically for `_simplex` is handled at the point of call. That way any other simplex solver could be implemented and the special handling would happen only at the point of call.

Reply:
> To make the reduction clear I think it should take place together, not split up over multiple functions.

This is the flow:
```
lpmin(*args)
    args -> interpreted into standard f, constr in _lp_args
        f, constr -> matrix with aux variables in _lp_matrices
            A,B,C,D -> solution' (perhaps in terms of aux) in _lp
                solution' -> final solution in lpmin
```

Reply:
Perhaps we make lpmin only accept f, constr. If the user wants to use matrices then the call would be 
* `lpmin(*primal_dual(M)[0])`  if M is for f and `Ax<=b` or 
* `lpmin(*primal_dual(M.T)[1])` if M is for f and `Ax>=b` or 
* `lpmin(*primal_dual(_m(a,b,c,d))[0])` if a,b,c,d are the parts from `f=cx-d` and `ax<=b`. 
* etc...

In other words, give them the tools for changing from one form to the other and let them use them or create lambda's as they desire.

Reply:
> If the user wants to use matrices

One of the reasons to use matrices is efficiency, so converting matrices to equations would defeat the purpose.